### PR TITLE
feat: Add suggestion controller

### PR DESCRIPTION
**Pull request by gaocegege**
Signed-off-by: Ce Gao <gaoce@caicloud.io>

**What this PR does / why we need it**:

Add the basic controller for suggestion CRD. It does not contain:

- Logic to create suggestions according to algorithms (only supports random now).
- Logic to communicate with suggestions (Use a fake implementation instead).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Ref #507 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/742)

### CODE REVIEW

File: pkg/controller.v1alpha3/suggestion/suggestion_controller.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		if errors.IsNotFound(err) {
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	oldS := instance.DeepCopy()
+
+	deploy, err := r.DesiredDeployment(instance)
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/suggestion_controller.go]: I am confused here. Why do we need a deployment and service for a suggestion?

File: pkg/controller.v1alpha3/suggestion/composer/composer.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	// Add owner reference to the service so that it could be GC after the suggestion is deleted
+	if err := controllerutil.SetControllerReference(s, service, g.scheme); err != nil {
+		return nil, err
+	}
+
+	return service, nil
+}
+
+// TODO(gaocegege): Implement switch logic.
+func (g *General) desiredContainer(s *suggestionsv1alpha3.Suggestion) *corev1.Container {
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/composer/composer.go]: why do  we need separate container per process?  Am I missing something?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller.go (hunk 1 above)
Comment (xieydd) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller.go]: I guess, suggestion controller will depend on the deployment controller and service controller, and binding the deployment and the service will expose a service port in cluster.
@gaocegege Can you correct my guess.

File: pkg/controller.v1alpha3/suggestion/composer/composer.go (hunk 2 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/composer/composer.go]: We need to create the algorithm container for the given algorithm name and settings.

File: pkg/controller.v1alpha3/suggestion/suggestion_controller.go (hunk 1 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller.go]: @johnugeorge I think we should create the suggestion first, then we can communicate with the suggestion.

For example, if an experiment uses random, we need to create a random suggestion for the experiment.

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		log.Info("Creating Deployment", "namespace", deploy.Namespace, "name", deploy.Name)
+		err = r.Create(context.TODO(), deploy)
+		return nil, err
+	} else if err != nil {
+		return nil, err
+	}
+	return foundDeploy, nil
+}
+
+func (r *ReconcileSuggestion) createOrUpdateService(service *corev1.Service) (*corev1.Service, error) {
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Aren't we updating the service here?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *ReconcileSuggestion) createOrUpdateDeployment(deploy *appsv1.Deployment) (*appsv1.Deployment, error) {
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Do we need to update the deployment if already present?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 3 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: No idea about if we have such a use case now.

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 4 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: No idea about if we have such a use case now.

File: pkg/controller.v1alpha3/suggestion/composer/composer.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	c := &corev1.Container{
+		Name: consts.ContainerSuggestion,
+	}
+	c.Image = "katib/v1alpha2/suggestion-random:1"
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/composer/composer.go]: Do you plan to add the mapping of algorithms in this PR?

Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/composer/composer.go]: No, I think we can do it in another PR.

File: pkg/controller.v1alpha3/suggestion/suggestion_controller.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+		return reconcile.Result{}, err
+	}
+
+	if err := r.List(context.TODO(),
+		client.MatchingLabels(util.TrialLabels(experiment)), trials); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	logger.Info("Sync assignments", "suggestions", instance.Spec.Suggestions)
+	if err = r.SyncAssignments(instance, experiment, trials.Items); err != nil {
```
Comment (johnugeorge) [inline pkg/controller.v1alpha3/suggestion/suggestion_controller.go]: Where is this function defined?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 4 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Ok. Function name is confusing. Either we can name to ReconcileDeployment?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 3 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Ok.  we can rename to ReconcileService?

File: pkg/controller.v1alpha3/suggestion/suggestion_controller.go (hunk 6 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller.go]: https://github.com/kubeflow/katib/pull/742/files#diff-1cdb066fdd3258bfce6a7218681d43b6R14

Now we only have a fake implementation. The real implementation should be added after GRPC API is updated. Ref https://github.com/kubeflow/katib/pull/735/files#diff-5ad38cda167bce34ad354bdef2f05d51R36

File: pkg/controller.v1alpha3/suggestion/composer/composer.go (hunk 5 above)
Comment (hougangliu) [reply pkg/controller.v1alpha3/suggestion/composer/composer.go]: I introduced katib-config as https://github.com/kubeflow/katib/pull/745/files#diff-916572ac262b489ac2debd624cb44116, you can add `suggestion` configuration in the configmap,  for each algorithm, except `image` field, maybe you can also add "is-long-running" field

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 3 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Sure

File: pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go (hunk 4 above)
Comment (gaocegege) [reply pkg/controller.v1alpha3/suggestion/suggestion_controller_util.go]: Sure

File: pkg/controller.v1alpha3/suggestion/composer/composer.go (hunk 5 above)
Comment (johnugeorge) [reply pkg/controller.v1alpha3/suggestion/composer/composer.go]: @hougangliu   Why not `image` field ? what is the structure you propose?

**Comment by johnugeorge** [comment]
Merging this PR 
/approve

**Pull request merged**